### ISSUE TITLE
What is gist's leading underscore naming convention (e.g., is it for NamedIndividuals)...?

### ISSUE THREAD

**Issue by pmcb55** [semanticarts/gist#967]
So I noticed a number of terms in gist core named with leading underscores, mostly base units (the only exception being `gist:_PrefixDeclaration_gist`):
```
gist:BaseUnit
	a owl:Class ;
	rdfs:subClassOf gist:SimpleUnitOfMeasure ;
	owl:equivalentClass [
		a owl:Class ;
		owl:oneOf (
			gist:_USDollar
			gist:_ampere
			gist:_bit
			gist:_candela
			gist:_each
			gist:_kelvin
			gist:_kilogram
			gist:_meter
			gist:_mole
			gist:_second
		) ;
	] ;
	skos:definition "A primitive unit that cannot be decomposed into other units. It can be converted from one measurement system to another.  The base units in gist are the seven primitive units from the System Internationale (SI): (meter, second, kilogram, ampere, kelvin, mole, candela), plus three convenience ones: each. bit and usDollar."^^xsd:string ;
	skos:prefLabel "Base Unit"^^xsd:string .
```
So I've a few questions:
1. Would it be correct to interpret this convention as: "**_Prefix all NamedIndividuals with a leading underscore_**" ?
2. If so, then is there any specific reason that each of these terms is not explicitly typed as an `owl:NamedIndividual`?
At the moment most are like this:
    ```
    gist:_ampere
	  a
		  owl:Thing ,
		  gist:BaseUnit
		  ;
    ```
   ...so why not this instead:
    ```
    gist:_ampere
	  a
		  owl:NamedIndividual ,
		  gist:BaseUnit
		  ;
    ```
3. It seems there is no strict convention for the case of the first letter (i.e., both `gist:_USDollar` and `gist:_PrefixDeclaration_gist` use an uppercase first letter, whereas all the rest use a lowercase first letter). Since NamedIndividuals are instances of a Class (and not a Property), I think I'd personally favour uppercasing the first letter (as everyone does today for Class terms). So then we'd have `gist:_Ampere`, `gist:_Second`, etc.
4. Has this convention raised any push-back from developers using these terms in their code (e.g., where they have static constants with names that reflect the gist term name, e.g., in Java something like:
    ```
    public class GIST { 
        public static final String _ampere = "https://w3id.org/semanticarts/ns/ontology/gist/_ampere";
    }
    ```
    ...because leading underscores can already have specific meanings in different programming language (e.g., in Python it means [these objects are used internally](https://towardsdatascience.com/whats-the-meaning-of-single-and-double-underscores-in-python-3d27d57d6bd1), in JavaScript and C/C++ it's a (not hugely common) convention that leading underscores denote private member variables, etc.), and in general these leading underscores might be frowned upon by developers seeing them in their code (e.g., violating there linting rules, or seen as un-idiomatic to their programming language, etc.) ...?
    
(Yes, generated, or manually written, source code like the above java example can always omit these leading underscores from the variable name, but then developers source code is slightly misaligned with the RDF, which I personally don't like.)

**Comment by rjyounes** [comment]
@pmcb55 Our best practice is to name individuals with leading underscore + name of most specific rigid class + underscore + name. That is, `gist:_PrefixDeclaration_gist` uses this standard, while `gist:_ampere` does not, merely for historical reasons - I.e., the convention was not articulated until after the gist unit of measure individuals were defined. See issue 5356 and the [gist style guide](https://github.com/semanticarts/gist/blob/develop/docs/gistStyleGuide.md#instance-local-names) for details. The relatively new unit of measure `gist:_percent` uses the older format for consistency with the other units.

We have not received any pushback from developers, and note that the gist individuals will never appear without a namespace, either prefix or full IRI, so there is no confusion with the programming naming convention. In addition, these are significantly different domains used in different contexts, and I don't think there's a likelihood that IRIs would be confused with method names.

**Comment by rjyounes** [comment]
Regarding typing of `owl:NamedIndividual`, it is redundant because the instances are always assigned a more specific type, and we asked the EDM Council, whose rdf-toolkit.jar serializer we use before repository commits, to provide an option to remove these. This is now the `-sni` or `--suppress-named-individuals` option; see [EDM's documentation of the tool](https://github.com/edmcouncil/rdf-toolkit). 

The use of the similarly redundant `owl:Thing` assertions is due to the current behavior of [`onto_tool`](https://github.com/semanticarts/ontology-toolkit), which we use for release packaging. See issues #916 and #775 (also yours!) for discussion.

I hope you enjoy history!

**Comment by pmcb55** [comment]
Ah ha - thanks @rjyounes.

Sorry to be so pedantic, but just to follow on from this: "_Regarding typing of `owl:NamedIndividual`, it is redundant because the instances are always assigned a more specific type_", looking at this example:
```
<https://w3id.org/semanticarts/ns/ontology/gist/_kilogram> a gist:BaseUnit, owl:Thing;
    skos:definition "The base unit for measuring mass.";
    gist:unitSymbol "kg".
```
...`gist:_kilogram` is defined as a `gist:BaseUnit`, so I thought, Ok, that must be explicitly typed as an `owl:NamedIndividual` then. But it's not, it's typed as an `owl:Class`:
```
gist:BaseUnit a owl:Class;
    owl:equivalentClass _:b29;
    rdfs:subClassOf gist:SimpleUnitOfMeasure.
```
...yet it's an `rdfs:subClassOf` `gist:SimpleUnitOfMeasure`, but that's also defined as just an `owl:Class`:
```
gist:SimpleUnitOfMeasure a owl:Class;
    skos:definition "Each simple unit has a base unit and...".
```

So, in other words, it seems now that nothing in gist is directly, _or indirectly_, declared to be an `owl:NamedIndividual` at all. Instead the actual named individuals (i.e., the terms with leading underscores) are all actually just sub-classes.

But aren't `owl:Class` (and sub-classes of classes) and `owl:NamedIndividual` very different things...?! If the distinction is important enough to provide a specific naming convention to help humans see that distinction (i.e., use a leading underscore, etc.), then isn't it also important for the machines to be able to make that same distinction too...!?

**Comment by uscholdm** [comment]
> Ok, so looking at this example:
> 
> ```
> <https://w3id.org/semanticarts/ns/ontology/gist/_kilogram> a gist:BaseUnit, owl:Thing;
>     skos:definition "The base unit for measuring mass.";
>     gist:unitSymbol "kg".
> ```
> 
> ...`gist:_kilogram` is defined as a `gist:BaseUnit`, so I thought, Ok, that must be explicitly typed as an `owl:NamedIndividual` then. But it's not, it's typed as an `owl:Class`:

In OWL-DL, all instances are linked to their class via `rdf:type`. So `gist:_kilogram` is an instance of the class, `BaseUnit` which is correct. By inference, any instance of any class is also an instance of  both `owl:NamedIndividual` and `owl:Thing`.  We don't want to clutter up gist with such assertions, which is why we remove them by setting the appropriate parameter when calling the serializer.

> Instead the actual named individuals (i.e., the terms with leading underscores) are all actually just sub-classes.

No, actual named individuals are _instances_ of classes or subclasses. `NamedIndividual` is a subclass of `owl:Thing`

**Comment by pmcb55** [comment]
> No, actual named individuals are instances of classes or subclasses. NamedIndividual is a subclass of owl:Thing

Yep, my mistake. Thanks for the clarification :smile: 

But just to help with my continuing education - would you ever see a justification (in gist, or any ontology) to explicitly type any instance using `owl:NamedIndividual` then...? In other words, what **_practical_** purpose might the term `owl:NamedIndividual` serve **_in general_**?

(And as an aside, would you consider it a more general Best Practice/Guideline that any-and-all 'instances' in any ontology should always be explicitly of some _'more specific type_' already (apart from `owl:Thing` I mean)?)

**Comment by uscholdm** [comment]
> And as an aside, would you consider it a more general Best Practice/Guideline that any-and-all 'instances' in any ontology should always be explicitly of some _'more specific type_' already (apart from `owl:Thing` I mean)?)

Yes, I can't think of a reason to create a direct instance of `owl:Thing`

**Comment by uscholdm** [comment]
@pmcb55
PROPOSAL: close this issue as there is no suggested action. It's been about clarification.

**Comment by pmcb55** [comment]
Thanks so much for the clarifications - I'm happy to close now.
